src/UserEscrow.sol

```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.18;
+
+import "./util/Auth.sol";
+
+interface TransferLike {
+    function transferFrom(address, address, uint256) external returns (bool);
+    function transfer(address, uint256) external returns (bool);
+}
+
+/**
+ * @dev Escrow contract that holds tokens for specific destinations.
+ * Ensures that once tokens are transferred in, they can only be
+ * transferred out to the pre-chosen destinations, by wards.
+ */
+contract UserEscrow is Auth {
+    event TransferIn(address indexed token, address indexed source, address indexed destination, uint256 amount);
+    event TransferOut(address indexed token, address indexed destination, uint256 amount);
+
+    mapping(address => mapping(address => uint256)) destinations; // map by token and destination
+
+    constructor() {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- Token approvals ---
+    function transferIn(address token, address source, address destination, uint256 amount) external auth {
+        require(TransferLike(token).transferFrom(source, address(this), amount), "UserEscrow/transfer-failed");
```

Comment:
if we also want to avoid transferFrom here, we need to call an auth transferOut function on the escrow